libs/MeshKernel/include/MeshKernel/UndoActions/UndoActionStack.hpp

```diff
@@ -49,31 +52,36 @@ namespace meshkernel
         // Perhaps Short for for menu items, long form for tooltips?
         // long form for any exceptions?
 
-        /// @brief Add an UndoAction.
+        /// @brief Add an UndoAction with an associated action-id.
         ///
         /// All added undo-actions must be in the committed state, if not then a ConstraintError
         /// will be raised.
         /// No null undo-actions will be added to the stack.
         /// All restored items will be removed, since after adding a new undo-action they are no
         /// longer restore-able.
-        void Add(UndoActionPtr&& transaction);
+        void Add(UndoActionPtr&& transaction, const int actionId = constants::missing::intValue);
 
         /// @brief Undo the action at the top of the committed stack
         ///
         /// The undo-action will be moved to the restored stack in case it should be re-done.
-        /// \returns true if an undo-action was performed, false otherwise
-        bool Undo();
+        /// \returns true if an undo-action was performed, false otherwise, and the associated action identifier.
+        /// If undo-action was false then the associated action identifier will be the value missing::intValue.
+        std::tuple<bool, int> Undo();
```

Comment:
one could consider using monads. https://www.modernescpp.com/index.php/monads-in-c/